Implement Module Toggling

BabyblueSheep opened the pull request:
Implements toggling modules with a JSON file, found in `tModLoader\CarbonMod\enabled-modules.json`.

While the method I used was a simple list of enabled modules, perhaps it could be better to have individual boolean values for each module?
Additionally, it might be better to add a value to check if an assembly is a module, instead of relying on the name alone.

BabyblueSheep committed:
create carbon folder

BabyblueSheep committed:
working configs

oli-x64 reviewed (changes requested):
The from-config loading system is good, and the file saves to the ideal location.

In terms of changes, we need to keep track of which assemblies are loaded for user configuration purposes, and there needs to be a method to save the current list of loaded assemblies to a file. For the second one you can probably just make a helper method out of part of the loading code that creates and populates the config file.

steviegt6 commented on src/Carbon/CarbonMod.cs:
```diff
@@ -23,6 +29,9 @@ namespace TeamCatalyst.Carbon
     ///     Carbon <see cref="Mod"/> entrypoint.
     /// </summary>
     public sealed class CarbonMod : Mod {
+        public static readonly string CarbonFolder = Path.Join(Main.SavePath, "CarbonMod");
+        public static readonly string ModuleConfigFile = Path.Join(Main.SavePath, "CarbonMod", "enabled-modules.json");
```
Prefer properties over resolving once; not a big deal, just preference.

steviegt6 commented on src/Carbon/CarbonMod.cs:
```diff
@@ -33,7 +42,7 @@ public override void Load()
         public override void Unload() {
             base.Unload();
-
+            
```
gang

steviegt6 commented on src/Carbon/CarbonMod.cs:
```diff
@@ -59,11 +73,50 @@ public override void Load()
+                    types.AddRange(pair.Value);
+                }
+                else
+                {
+                    if (names.Contains(assemblyName))
+                    {
+                        types.AddRange(pair.Value);
+                    }
```
Do we automatically handle enabling newly-introduced modules?

BabyblueSheep committed:
rewrite of module config

BabyblueSheep committed:
oops forgot to erase removed property

BabyblueSheep committed:
pre-build event that automatically adds dll refs

BabyblueSheep committed:
nvm forgot to remove debug echo

BabyblueSheep commented:
Writing the changes I've made for clarity:
- changed the `enabled-modules.json`; it now has a boolean for each module. This is so that newly added modules are automatically enabled
- added a custom assembly attribute for checking if an assembly is a module
- the mod now stores a dictionary of enabled modules and a list of loaded assemblies
- added a pre-build script that puts every referenced `.dll` in the `build.txt`
